src/NHibernate/Type/AbstractDateTimeType.cs

```diff
@@ -0,0 +1,160 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Globalization;
+using NHibernate.Engine;
+using NHibernate.SqlTypes;
+
+namespace NHibernate.Type
+{
+	/// <summary>
+	/// Base class for date time types.
+	/// </summary>
+	[Serializable]
+	public abstract partial class AbstractDateTimeType : PrimitiveType, IIdentifierType, ILiteralType, IVersionType
+	{
+		private static readonly DateTime _baseDateValue = DateTime.MinValue;
+
+		/// <summary>
+		/// Returns the <see cref="DateTimeKind" /> for the type.
+		/// </summary>
+		protected virtual DateTimeKind Kind => DateTimeKind.Unspecified;
+
+		/// <inheritdoc />
+		public override System.Type ReturnedClass => typeof(DateTime);
+
+		/// <summary>
+		/// Retrieve the current system time.
+		/// </summary>
+		/// <value><see cref="DateTime.UtcNow" /> if <see cref="Kind" /> is <see cref="DateTimeKind.Utc" />,
+		/// <see cref="DateTime.Now" /> otherwise.</value>
+		protected virtual DateTime Now => Kind == DateTimeKind.Utc ? DateTime.UtcNow : DateTime.Now;
+
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		protected AbstractDateTimeType() : base(SqlTypeFactory.DateTime)
+		{
+		}
+
+		/// <summary>
+		/// Constructor for overriding the default <see cref="SqlType"/>.
+		/// </summary>
+		/// <param name="sqlTypeDateTime">The <see cref="SqlType"/> to use.</param>
+		protected AbstractDateTimeType(SqlType sqlTypeDateTime) : base(sqlTypeDateTime)
+		{
+		}
+
+		/// <summary>
+		/// Adjust the date time value for this type from an arbitrary date time value.
+		/// </summary>
+		/// <param name="dateValue">The <see cref="System.DateTime" /> to adjust.</param>
+		/// <returns>A <see cref="System.DateTime" />.</returns>
+		protected virtual DateTime AdjustDateTime(DateTime dateValue) =>
+			Kind == DateTimeKind.Unspecified ? dateValue : DateTime.SpecifyKind(dateValue, Kind);
```

Comment:
From `AbstractDateTimeSpecificKindType`, but applied only when specified. This method `AdjustDateTime` take the purpose of `CreateDateTime` which was in `DateTimeType`.